File: src/yadism/coefficient_functions/light/n3lo/xclsg3p.py
import numba as nb
import numpy as np

from .common import d27, d81


@nb.njit("f8(f8,f8[:])", cache=True)
def cls3a(y, args):
    nf = args[0]
    flps = args[1]
    dl = np.log(y)
    y1 = 1.0 - y
    dl1 = np.log(1.0 - y)

    cls31 = (
        (1568.0 * d27 * dl1**3 - 11904.0 * d27 * dl1**2 + 5124.0 * dl1) * y1**2
        + dl * dl1 * (2184.0 * dl + 6059.0 * y1)
        - (795.6 + 1036.0 * y) * y1**2
        - 143.6 * dl * y1
        + 8544.0 * d27 * dl**2
        - 1600.0 * d27 * dl**3
        - 885.53 / y * y1**2
        - 182.00 * dl / y * y1
    )
    cls32 = (
        (-96.0 * d27 * dl1**2 + 29.52 * dl1) * y1**2
        + +dl * dl1 * (35.18 * dl + 73.06 * y1)
        - (14.16 - 69.84 * y) * y1**2
        - 35.24 * y * dl**2
        - 69.41 * dl * y1
        - 384.0 * d27 * dl**2
        + 40.239 / y * y1**2
    )
    cls3F = (
        (107.0 + 321.05 * y - 54.62 * y**2) * (1.0 - y)
        - 26.717
        - 320.0 * d81 * dl**3
        - 640.0 * d81 * dl**2
        + 9.773 * dl
        + y * dl * (363.8 + 68.32 * dl)
    ) * y
    # Note here the source file cointain a typo and the
    # proper color factor is flps = fls - fl, not just fls.
    # see https://arxiv.org/pdf/hep-ph/0411112.pdf eq 9.
    res = nf * (cls31 + (flps) * cls3F + nf * cls32)
    return res


@nb.njit("f8(f8,f8[:])", cache=True)
def clg3a(y, args):
    nf = args[0]
    flg = args[1]
    dl = np.log(y)
    y1 = 1.0 - y
    dl1 = np.log(1.0 - y)

    clg31 = (
        (
            144.0 * dl1**4
            - 47024.0 * d27 * dl1**3
            + 6319.0 * dl1**2
            + 53160.0 * dl1
        )
        * y1
        + dl * dl1 * (72549.0 + 88238.0 * dl)
        + (3709.0 - 33514.0 * y - 9533.0 * y**2) * y1
        + 66773.0 * y * dl**2
        - 1117.0 * dl
        + 45.37 * dl**2
        - 5360.0 * d27 * dl**3
        - 2044.70 / y * y1
        - 409.506 * dl / y
    )
    clg32 = (
        (
            288.0 * d27 * dl1**3
            - 3648.0 * d27 * dl1**2
            - 592.3 * dl1
            + 1511.0 * y * dl1
        )
        * y1
        + dl * dl1 * (311.3 + 14.24 * dl)
        + (577.3 - 729.0 * y) * y1
        + 30.78 * y * dl**3
        + 366.0 * dl
        + 3000.0 * d27 * dl**2
        + 480.0 * d27 * dl**3
        + 88.5037 / y * y1
    )
    clg3F = (
        (
            -0.0105 * dl1**3
            + 1.550 * dl1**2
            + 19.72 * y * dl1
            - 66.745 * y
            + 0.615 * y**2
        )
        * y1
        + 20.0 * d27 * y * dl**4
        + (280.0 / 81.0 + 2.260 * y) * y * dl**3
        - (15.40 - 2.201 * y) * y * dl**2
        - (71.66 - 0.121 * y) * y * dl
    )
    res = nf * (clg31 + nf * (clg32 + flg * clg3F))
    return res

File: extras/light/n3lo_source/xclsg3p.f
*
* ..File: xclsg3p.f    F_L^PS  and  F_L^G
*
*
* ..Parametrizations of the 3-loop MS(bar) pure-singlet and gluon coef-
*    ficient functions for the electromagnetic structure function F_L
*    at  mu_r = mu_f = Q. The expansion parameter is  alpha_s/(4 pi).
*
*  ..The relative accuracy of these parametrizations, as well as of
*    the convolution results, is one part in thousand or better.
*
* ..Reference: S. Moch, J. Vermaseren and A. Vogt,
*              hep-ph/0411112 = Phys. Lett. B606 (2005) 123
*
* =====================================================================
*
*
* ..The pure-singlet coefficient function
*
       FUNCTION CLS3A (Y, NF)
       IMPLICIT REAL*8 (A-Z)
       DIMENSION FLS(6)
       INTEGER NF
       DATA FLS / 1.d0, 0.1d0, 0.d0, 0.1d0, 0.01818181818d0, 0.1d0 /
*
       DL  = LOG (Y)
       Y1  = 1.-Y
       DL1 = LOG (Y1)
       D27 = 1./27.D0
       D81 = 1./81.D0
*
       CLS31 = (1568.*D27 * DL1**3 - 11904.*D27 * DL1**2 + 5124.* DL1)
     ,           * Y1**2  +  DL*DL1 * (2184.* DL + 6059.* Y1)
     ,         - (795.6 + 1036.* Y) * Y1**2  - 143.6 * DL*Y1
     ,         + 8544.*D27 * DL**2 - 1600.*D27 * DL**3
     ,         - 885.53 /Y *Y1**2 - 182.00 * DL/Y * Y1
       CLS32 = ( - 96.*D27 * DL1**2 + 29.52 * DL1) * Y1**2  +
     ,         +  DL*DL1 * (35.18 * DL + 73.06 * Y1)
     ,         - (14.16 - 69.84 * Y) * Y1**2 - 35.24 * Y*DL**2
     ,         - 69.41 * DL*Y1 - 384.*D27 * DL**2 + 40.239 /Y *Y1**2
       CLS3F = ( (107.0 + 321.05 * Y - 54.62 * Y**2) *(1.-Y)
     ,         - 26.717 - 320.*D81 * DL**3 - 640.*D81 * DL**2
     ,         + 9.773 * DL + Y*DL * (363.8 + 68.32 * DL) ) * Y
       CLS3A = NF * ( CLS31 + FLS(NF) * CLS3F + NF * CLS32 )
*
       RETURN
       END
*
* ---------------------------------------------------------------------
*
*
* ..The gluon coefficient function
*
       FUNCTION CLG3A (Y, NF)
       IMPLICIT REAL*8 (A-Z)
       DIMENSION FLG(6)
       INTEGER NF
       DATA FLG / 1.d0, 0.1d0, 0.d0, 0.1d0, 0.01818181818d0, 0.1d0 /
*
       DL  = LOG (Y)
       Y1  = 1.-Y
       DL1 = LOG (Y1)
       D27 = 1./27.D0
*
       CLG31 = (144.D0 * DL1**4 - 47024.*D27 * DL1**3 + 6319.* DL1**2
     ,         + 53160.* DL1) * Y1 + DL*DL1 * (72549. + 88238.* DL)
     ,         + (3709. - 33514.* Y - 9533.* Y**2) * Y1
     ,         + 66773.* Y*DL**2 - 1117.* DL + 45.37 * DL**2
     ,         - 5360.*D27 * DL**3 - 2044.70 /Y*Y1 - 409.506 * DL/Y
       CLG32 = (288.*D27 * DL1**3 - 3648.*D27 * DL1**2 - 592.3 * DL1
     ,         + 1511.* Y*DL1) * Y1 + DL*DL1 * (311.3 + 14.24 * DL)
     ,         + (577.3 - 729.0 * Y) * Y1 + 30.78 * Y*DL**3
     ,         + 366.0 * DL + 3000.*D27 * DL**2 + 480.*D27 * DL**3
     ,         + 88.5037 /Y*Y1
       CLG3F = (-0.0105 * DL1**3 + 1.550 * DL1**2 + 19.72 *Y*DL1
     ,         - 66.745 * Y + 0.615 * Y**2) * Y1 + 20.*D27 * Y*DL**4
     ,         + (280./81.D0 + 2.260* Y) * Y*DL**3 - (15.40 - 2.201* Y)
     ,           * Y*DL**2 - (71.66 - 0.121 * Y) * Y*DL
       CLG3A = NF * ( CLG31 + NF * (CLG32 + FLG(NF) * CLG3F) )
*
       RETURN
       END
*
* =================================================================av==
